frontend/src/metabase-lib/parameters/utils/click-behavior.ts

```diff
@@ -124,75 +127,138 @@ export function getDataFromClicked({
   return { column, parameter, parameterByName, parameterBySlug, userAttribute };
 }
 
-const { Text, Number, Temporal } = TYPE;
-
 function notRelativeDateOrRange({ type }: Parameter) {
   return type !== "date/range" && type !== "date/relative";
 }
 
 export function getTargetsForQuestion(question: Question): Target[] {
-  const query = question.legacyQuery({ useStructuredQuery: true });
-  return [...query.dimensionOptions().all(), ...query.variables()].map(o => {
-    let id, target: ClickBehaviorTarget;
-    if (o instanceof TemplateTagVariable || o instanceof TemplateTagDimension) {
-      let name;
-      ({ id, name } = o.tag());
-      target = { type: "variable", id: name };
-    } else if ("mbql" in o) {
-      const dimension: ClickBehaviorDimensionTarget["dimension"] = [
-        "dimension",
-        o.mbql(),
-      ];
-      id = JSON.stringify(dimension);
-      target = { type: "dimension", id, dimension };
-    } else {
-      throw new Error("Unknown target type");
-    }
-    let parentType: string | undefined | null;
-    let parameterSourceFilter: SourceFilters["parameter"] = () => true;
-    if (o instanceof TemplateTagVariable) {
-      const type = o.tag()?.type;
-      parentType = type
-        ? {
-            card: undefined,
-            dimension: undefined,
-            snippet: undefined,
-            text: Text,
-            number: Number,
-            date: Temporal,
-          }[type]
-        : undefined;
-      parameterSourceFilter = parameter =>
-        variableFilterForParameter(parameter)(o);
-    } else {
-      const field = o.field();
-
-      if (field != null) {
-        const { base_type } = field;
-        parentType =
-          [Temporal, Number, Text].find(
-            t => typeof base_type === "string" && isa(base_type, t),
-          ) || base_type;
-        parameterSourceFilter = parameter =>
-          dimensionFilterForParameter(parameter)(o);
-      }
-    }
+  if (question.isStructured()) {
+    return getTargetsForStructuredQuestion(question);
+  }
+
+  if (question.isNative()) {
+    return getTargetsForNativeQuestion(question);
+  }
+
+  return [];
+}
+
+function getTargetsForStructuredQuestion(question: Question): Target[] {
+  const query = question.query();
+  const stageIndex = -1;
+  const visibleColumns = Lib.visibleColumns(query, stageIndex);
+
+  return visibleColumns.map(targetColumn => {
+    const dimension: ClickBehaviorDimensionTarget["dimension"] = [
+      "dimension",
+      Lib.legacyRef(targetColumn),
+    ];
+    const id = JSON.stringify(dimension);
+    const target: ClickBehaviorTarget = { type: "dimension", id, dimension };
 
     return {
       id,
       target,
-      name: o.displayName({ includeTable: true }),
+      name: Lib.displayInfo(query, stageIndex, targetColumn).longDisplayName,
       sourceFilters: {
-        column: column =>
+        column: (sourceColumn, sourceQuestion) => {
+          if (!sourceQuestion) {
```

Comment:
we pass `question` as `query` from the closure is becoming outdated

Reply:
If `sourceQuestion` is required this won't be needed